### ISSUE TITLE
Cleanup of the reconcile loop after #4086

### ISSUE THREAD

**Issue by Skarlso** [external-secrets/external-secrets#4142]
There are a number of points in that PR that we will leave for a followup issue:

- Complexity score
- https://github.com/external-secrets/external-secrets/pull/4086#discussion_r1854867539
- https://github.com/external-secrets/external-secrets/pull/4086#discussion_r1855520816

**Comment by thesuperzapper** [comment]
Also, we should follow up to see if the check we removed that was added in https://github.com/external-secrets/external-secrets/pull/3746 should be re-implemented in a way that it actually does something.

I removed it because it did not do anything, search the diff in https://github.com/external-secrets/external-secrets/pull/4086 for `if errors.Is(err, esv1beta1.NotModifiedErr) {` in the ES reconcile function to see its code.

**Comment by Skarlso** [comment]
Oh I thought you _fixed_ that not just removed it. :D

Ok, yeah, that should do something. The interesting part is that people reported that it works now. :D

**Comment by thesuperzapper** [comment]
Another "cleanup" task which should reduce the complexity of `Reconcile()` is to simply move the `mutationFunc := func(secret *v1.Secret) error {` definition into a method on the `Reconciler` object, or possibly a function if people prefer.

Additionally, it's a bit dangerous that we currently define it in-line, because developers might accidentally use the `existingSecret` object, rather than the `secret` object that it is passed.

**Comment by Skarlso** [comment]
I absolutely 💯 agree. That function should pretty much live outside by now.

**Comment by thesuperzapper** [comment]
Also, I feel like we should move this section of `mutationFunc` to the `r.applyTemplate()` method:

https://github.com/external-secrets/external-secrets/blob/3218a21cb7e7905137e947e3166f52c30f4b7c6d/pkg/controllers/externalsecret/externalsecret_controller.go#L409-L419

Because we already do a similiar thing within `applyTemplate()` to remove labels/annotations we manage:

https://github.com/external-secrets/external-secrets/blob/3218a21cb7e7905137e947e3166f52c30f4b7c6d/pkg/controllers/externalsecret/externalsecret_controller_template.go#L113-L128

**Comment by evrardj-roche** [comment]
@Skarlso the code has moved a lot since then, any opinion on whether this is still worth pursuing?

**Comment by Skarlso** [comment]
I mean it's sort of related to what you are trying to do with the extraction, right?

**Comment by evrardj-roche** [comment]
Completely :) 

Allow me to clarify my thinking.

FYI I will not work on my design proposal of extraction until we know and agree on the trajectory (proposal has been merged).
I already have been through some large efforts to show some PoC, which resulted in no merge in either the effort or the proposal... I won't do that mistake twice :) 

Keep in mind my ideas were just that, ideas. You already had some ideas too.
So instead of me going through the ropes without learning from you, I wanted to know if this is still valid and how we can improve together. I could be the hands to your brain!

When I have written the controller separation design proposal, I realized well that some parts would be hard. Those _can be done through steps_, steps who have been already _walked_ by the community (example, @gusfcarvalho code refactoring a bit of the Reconcile loop in one of his work for ESI). If this issue _is a step_, why not pursuing it, in the large scheme of things?

I am here to help, tell me how we're gonna wing it.

**Comment by Skarlso** [comment]
Ok then.
